### PR TITLE
Redo some indexing inside FESystem.

### PR DESCRIPTION
**Pull request by drwells**
This function takes up about 3% of the total runtime for an application I'm working on. We can reduce that cost by nearly 50% by moving some checks outside of loops and explicitly using `std::copy()`, which ultimately results in the compiler calling `memcpy()` (it can't do that on the present version due to aliasing).

### CODE REVIEW

File: source/fe/fe_system.cc [hunk 1]
```diff
@@ -147,31 +147,39 @@ namespace internal
...
-            for (unsigned int q = 0; q < n_q_points; ++q)
-              output_data.shape_hessians[out_index][q] =
-                base_data.shape_hessians[b][q];
-
-          if (base_flags & update_3rd_derivatives)
-            for (unsigned int q = 0; q < n_q_points; ++q)
-              output_data.shape_3rd_derivatives[out_index][q] =
-                base_data.shape_3rd_derivatives[b][q];
-        }
+    auto copy_row = [&](const auto row_in, auto row_out) {
```
Comment (bangerth) [inline source/fe/fe_system.cc:151]: The only thing you're capturing here is `n_q_point`. Would you be willing to make that explicit?
```suggestion
    auto copy_row = [n_q_point](const auto row_in, auto row_out) {
```

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
OK otherwise!

**Review by kronbichler — APPROVED** [review]
OK for me as well.

File: source/fe/fe_system.cc (hunk 1 above)
Comment (drwells) [reply source/fe/fe_system.cc]: We don't need `n_q_points` at all since we can just use the end iterator (which turns out to be slightly faster anyway).

**Review by kronbichler — APPROVED** [review]

**Pull request merged**